src/dodal/utils.py

```diff
@@ -110,6 +114,68 @@ def wrapper(*args, **kwds) -> T:
     return decorator
 
 
+class DeviceInitializationController(Generic[D]):
+    def __init__(
+        self,
+        factory: Callable[[], D],
+        eager_connect: bool,
+        use_factory_name: bool,
+        timeout: float,
+        mock: bool,
+        skip: skip,
+    ):
+        self._factory: Callable[[], D] = factory
+        self._cached_device: D | None = None
+        self._callbacks: list[Callable[[D], None]] = []
+        self._eager_connect = eager_connect
+        self._use_factory_name = use_factory_name
+        self._timeout = timeout
+        self._mock = mock
+        self._skip = skip
+        update_wrapper(self, factory)
+
+    @property
+    def skip(self) -> bool:
+        return self._skip() if callable(self._skip) else self._skip
+
+    @property
+    def device(self) -> D | None:
+        return self._cached_device
+
+    def add_callback(self, callback: Callable[[D], None]):
+        self._callbacks.append(callback)
+
+    def __call__(
+        self,
+        connect: bool | None = None,
+        name: str | None = None,
+        timeout: float | None = None,
+        mock: bool | None = None,
```

Comment:
Could: This is basically the interface for when I do `i22.dcm(...)` from a script. As such it would be better to make the arg names a bit cleaner e.g. `connect_immediately`? and `connection_timeout`

Reply:
Also docstrings but happy for that to be covered by https://github.com/DiamondLightSource/dodal/issues/780

Reply:
As an aside: I've been in enough debates about verbosity vs. "why say many word when few do trick" now to wonder if we need a style guide. Java is very pro-verobosity and has that culture around it so we never needed one with GDA, the Python world is a bit more fragmented.

Reply:
Ooof, it'll be controversial and I think hard to codify into a style guide. I feel like the answer is always "As verbose as is required to make it clear given the context" which is so subjective.  But I would be happy to weigh in on a discussion

Reply:
It is quite subjective, but it's actually quite useful. What you said is a nice middle ground between "you should minimise the amount of lines/characters you write" (C) or "You should write function signatures as full sentences" (Java)

Reply:
:sunglasses: :+1: